File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    dummy->val = -1;  
    dummy->next = NULL;
    struct ListNode* head = dummy;
    while(list1 && list2){
        if(list1->val <= list2->val){
            dummy->next = list1; 
            list1 = list1->next; 
        }
        else{
            dummy->next = list2; 
            list2 = list2->next; 
        }
        dummy = dummy->next; 
    }
    while(list1){
        dummy->next = list1; 
        list1 = list1->next;  
        dummy = dummy->next; 
    }
    while(list2){
        dummy->next = list2; 
        list2 = list2->next; 
        dummy = dummy->next; 
    }
    return head->next; 
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>,
        mut list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {

        let mut dummy = Box::new(ListNode { val: 0, next: None });
        let mut tail = &mut dummy;

        while list1.is_some() && list2.is_some() {
            let l1 = list1.as_mut().unwrap();
            let l2 = list2.as_mut().unwrap();

            if l1.val <= l2.val {
                let next = l1.next.take();
                tail.next = list1.take();
                list1 = next;
            } else {
                let next = l2.next.take();
                tail.next = list2.take();
                list2 = next;
            }

            tail = tail.next.as_mut().unwrap();
        }

        tail.next = if list1.is_some() { list1 } else { list2 };

        dummy.next
    }
}
